File: ch08/reactive-functional-endpoints/course-tracker-api/src/main/java/com/manning/sbip/ch08/component/CourseHandler.java
package com.manning.sbip.ch08.component;

import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.web.reactive.function.BodyInserters.fromValue;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;

import com.manning.sbip.ch08.model.Course;
import com.manning.sbip.ch08.repository.CourseRepository;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Component
public class CourseHandler {

	private CourseRepository courseRepository;

	@Autowired
	public CourseHandler(CourseRepository courseRepository) {
		this.courseRepository = courseRepository;
	}

	public Mono<ServerResponse> findAllCourses(ServerRequest serverRequest) {
		Flux<Course> courses = this.courseRepository.findAll();
		return ServerResponse.ok().contentType(APPLICATION_JSON).body(courses, Course.class);
	}

	public Mono<ServerResponse> findCourseById(ServerRequest serverRequest) {
		String courseId = serverRequest.pathVariable("id");
		Mono<Course> courseMono = this.courseRepository.findById(courseId);
		return courseMono.flatMap(course -> ServerResponse.ok().contentType(APPLICATION_JSON).body(fromValue(course)))
				.switchIfEmpty(notFound());
	}

	public Mono<ServerResponse> createCourse(ServerRequest serverRequest) {
		Mono<Course> courseMono = serverRequest.bodyToMono(Course.class);

		return courseMono.flatMap(course -> ServerResponse.status(HttpStatus.CREATED).contentType(APPLICATION_JSON)
				.body(this.courseRepository.save(course), Course.class));
	}

	public Mono<ServerResponse> updateCourse(ServerRequest serverRequest) {
		String courseId = serverRequest.pathVariable("id");
		Mono<Course> existingCourseMono = this.courseRepository.findById(courseId);
		Mono<Course> newCourseMono = serverRequest.bodyToMono(Course.class);
		return newCourseMono
				.zipWith(existingCourseMono,
						(newCourse, existingCourse) -> Course.builder().id(existingCourse.getId())
								.name(newCourse.getName()).category(newCourse.getCategory())
								.rating(newCourse.getRating()).description(newCourse.getDescription()).build())
				.flatMap(course -> ServerResponse.ok().contentType(APPLICATION_JSON)
						.body(this.courseRepository.save(course), Course.class))
				.switchIfEmpty(notFound());
	}

	public Mono<ServerResponse> deleteCourse(ServerRequest serverRequest) {
		String courseId = serverRequest.pathVariable("id");
		return this.courseRepository.findById(courseId)
				.flatMap(existingCourse -> ServerResponse.ok().build(this.courseRepository.deleteById(courseId)))
				.switchIfEmpty(notFound());
	}

	public Mono<ServerResponse> deleteAllCourses(ServerRequest serverRequest) {
		return ServerResponse.ok().build(this.courseRepository.deleteAll());
	}

	private Mono<ServerResponse> notFound() {
		return ServerResponse.notFound().build();
	}

}

File: ch10/Kotlin/course-tracker-kotlin-rest-api/src/main/kotlin/com/manning/sbip/ch10/service/CourseHandler.kt
package com.manning.sbip.ch10.service

import com.manning.sbip.ch10.exception.CourseNotFoundException
import com.manning.sbip.ch10.model.Course
import com.manning.sbip.ch10.repository.CourseRepository
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class CourseHandler (private val courseRepository: CourseRepository) {

    fun createCourse(course: Course): Course =  courseRepository.save(course)

    fun createBulkCourses(courses: List<Course>): Iterable<Course> =  courseRepository.saveAll(courses)

    fun findCourseById(courseId: Long): Course = courseRepository.findById(courseId)
            .orElseThrow { CourseNotFoundException(HttpStatus.NOT_FOUND, "No course with supplied course id was found") }

    fun findAllCourses(): Iterable<Course> =  courseRepository.findAll()

    fun updateCourse(courseId: Long, updatedCourse: Course): Course  {
        return if(courseRepository.existsById(courseId)) {
            courseRepository.save(
                    Course(
                            id = updatedCourse.id,
                            name = updatedCourse.name,
                            category = updatedCourse.category,
                            rating = updatedCourse.rating,
                            description = updatedCourse.description
                    )
            )
        }
        else throw CourseNotFoundException(HttpStatus.NOT_FOUND, "No course with supplied course id was found")
    }


    fun deleteCourseById(courseId: Long) {
        return if (courseRepository.existsById(courseId)) {
            courseRepository.deleteById(courseId)
        } else throw CourseNotFoundException(HttpStatus.NOT_FOUND, "No course with supplied course id was found")
    }
}